Async close of `IndexShard`

DaveCTurner opened the pull request:
Very early WIP mainly to get some CI cycles

Relates #89821
Relates #107513
Relates ES-8334

DaveCTurner committed:
Extract IndexShardTestCase#closeShard

DaveCTurner committed:
WIP

DaveCTurner committed:
WIP

DaveCTurner committed:
WIP

DaveCTurner committed:
WIP

DaveCTurner committed:
WIP

DaveCTurner committed:
WIP

DaveCTurner committed:
WIP

DaveCTurner committed:
WIP

DaveCTurner committed:
Comments

DaveCTurner committed:
WIP

DaveCTurner committed:
WIP

DaveCTurner committed:
WIP

DaveCTurner committed:
WIP

DaveCTurner committed:
WIP

DaveCTurner committed:
WIP

DaveCTurner committed:
WIP

DaveCTurner committed:
WIP

DaveCTurner committed:
WIP

DaveCTurner committed:
WIP

DaveCTurner committed:
WIP

DaveCTurner committed:
WIP

DaveCTurner committed:
Actually fork

DaveCTurner committed:
Iter

DaveCTurner committed:
Merge remote-tracking branch 'upstream/main' into 2024/04/25/IndexShard-async-close

DaveCTurner committed:
closeShardNoCheck

DaveCTurner committed:
WIP

DaveCTurner committed:
WIP

DaveCTurner committed:
Comments

DaveCTurner committed:
Fix leaks

DaveCTurner committed:
Merge branch 'main' into 2024/04/25/IndexShard-async-close

DaveCTurner committed:
Merge remote-tracking branch 'upstream/main' into 2024/04/25/IndexShard-async-close

DaveCTurner committed:
Add testAsyncCloseShard

henningandersen commented on server/src/main/java/org/elasticsearch/indices/IndicesService.java:
```diff
@@ -915,17 +934,22 @@ public void removeIndex(final Index index, final IndexRemovalReason reason, fina
+            indexService.close(
+                extraInfo,
+                reason == IndexRemovalReason.DELETED,
+                EsExecutors.DIRECT_EXECUTOR_SERVICE,
```
Should this use the `shardCloseExecutor` instead?

```suggestion
                shardCloseExecutor,
```

DaveCTurner replied:
Ah thanks yes I missed that one, see e81ea9015cc

DaveCTurner replied:
... which of course was the one which broke all the things 🤦

henningandersen commented on server/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java:
```diff
@@ -190,6 +193,9 @@ public IndicesClusterStateService(
         this.shardLockRetryInterval = SHARD_LOCK_RETRY_INTERVAL_SETTING.get(settings);
         this.shardLockRetryTimeout = SHARD_LOCK_RETRY_TIMEOUT_SETTING.get(settings);
+
+        // ES-8334 TBD generic is probably wrong for this, we don't want massive parallelism. Maybe MANAGEMENT? Or something throttled?
```
I think the main risk here is that the pool is completely depleted by these tasks and thus I'd be in favor of throttling it or a dedicated pool. These threads will basically be just waiting for other work to complete, so a relatively high thread count seems appropriate (10x processors or something like that, unlike MANAGEMENT which has max 5).

I wonder if we'd want to eventually push back on the cluster applier thread if it fills up? Or if we prefer to take that risk and ultimately OOM on it? Hopefully closing will soon be quick(er).

DaveCTurner replied:
It's a little tricky because we might be doing IO on these threads (e.g. flushing) - they're only going to be waiting in the unlucky case where the `IndexWriter` is in the uncancellable step of a merge. So ideally we'd want enough threads to (probably) avoid them all hitting the unlucky case at the same time, without so many that we consume all IO on the node.

My intuition here is that we won't see very many of these tasks in nodes which are running normally, there's not that many shards moving off such nodes, and on a shutting-down node we won't be starting up any new shards so the number of these tasks is bounded by the number of shards to close. The bad case would be a normally-running node with very high churn in shards, starting up new shards so fast that it can't close the old ones down fast enough. Maybe we could block or throttle new shards starting while old shards are still shutting down, given that starting new shards is already async? Still this seems like an unlikely situation, I'd prefer to defer it until later. I'll see if I can track the size of the problem in a metric somehow.

henningandersen replied:
Thanks, makes sense, I'd obviously be ok with relatively few threads for this (given that we only have one today) as a starting point.

henningandersen reviewed (commented):
Took a preliminary look, looks good.

For the remaining TBDs, I think blocking on the failure paths is fine for this PR, I think those are essentially left as is, i.e., we've not introduced any new problems.

DaveCTurner committed:
Wrong executor

DaveCTurner committed:
Spotless

DaveCTurner committed:
Attempt at assertions

DaveCTurner committed:
Flush pool to complete test

DaveCTurner committed:
Fix TODO

DaveCTurner committed:
Delay processPendingDeletes

DaveCTurner committed:
Await in SearchServiceTests too

DaveCTurner committed:
Delay shard delete until after close

DaveCTurner committed:
Comment

DaveCTurner committed:
Merge branch 'main' into 2024/04/25/IndexShard-async-close

DaveCTurner committed:
Comment

DaveCTurner committed:
More await

DaveCTurner committed:
Wait for earlier cluster states too

DaveCTurner committed:
Introduce async closing of `IndexShard`

Closing an `IndexShard` may be expensive if the `IndexWriter` takes a
long time to roll back and shut down. This commit adds the facility for
forking the expensive work onto a different thread, although in
production it still runs this work directly.

Relates ES-8334

DaveCTurner committed:
Merge branch 'main' into 2024/04/25/IndexShard-async-close

DaveCTurner committed:
Merge branch 'main' into 2024/04/25/IndexShard-async-close

DaveCTurner committed:
Merge remote-tracking branch 'upstream/main' into 2024/04/25/IndexShard-async-close

DaveCTurner committed:
Moar async

DaveCTurner committed:
Merge branch 'main' into 2024/04/30/IndexShard-async-shard-close-part-1

DaveCTurner committed:
Comments

DaveCTurner committed:
Merge branch 'main' into 2024/04/30/IndexShard-async-shard-close-part-1

DaveCTurner committed:
Ensure listener completed

DaveCTurner committed:
Merge branch 'main' into 2024/04/25/IndexShard-async-close

DaveCTurner committed:
Fix IndexShardIT

DaveCTurner committed:
Merge branch 'main' into 2024/04/25/IndexShard-async-close

DaveCTurner committed:
Merge branch 'main' into 2024/04/30/IndexShard-async-shard-close-part-1

DaveCTurner committed:
Merge branch 'main' into 2024/04/25/IndexShard-async-close

DaveCTurner committed:
Wait before checking breakers

DaveCTurner committed:
Merge branch 'main' into 2024/04/30/IndexShard-async-shard-close-part-1

DaveCTurner committed:
Merge branch '2024/04/30/IndexShard-async-shard-close-part-1' into 2024/04/25/IndexShard-async-close

DaveCTurner committed:
Fix leaks

DaveCTurner committed:
Merge branch '2024/04/30/IndexShard-async-shard-close-part-1' into 2024/04/25/IndexShard-async-close

DaveCTurner committed:
WIP

DaveCTurner committed:
Merge branch 'main' into 2024/04/25/IndexShard-async-close

DaveCTurner committed:
executeDirectly

DaveCTurner committed:
More cleanup

DaveCTurner committed:
Revert & comment

DaveCTurner commented:
Seems good to go now, I squashed it down to one commit and opened https://github.com/elastic/elasticsearch/pull/108145; closing this.
